### PR TITLE
Add decorator for Sentry tracing

### PR DESCRIPTION
**Pull request by ynouri**
Draft proposal for https://github.com/getsentry/sentry-python/issues/760

### CODE REVIEW

**Comment by ynouri** [comment]
> This pull request has gone three weeks without activity. In another week, I will close it.
> 
> But! If you comment or otherwise update it, I will reset the clock, and if you label it `Status: Backlog` or `Status: In Progress`, I will leave it alone ... forever!
> 
> ----
> 
> "A weed is but an unloved flower." ― _Ella Wheeler Wilcox_ 🥀

I would be happy to bring this to the finish line but probably best to get feedback from a maintainer first. Thank you

File: sentry_sdk/decorators.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    Usage:
+    @sentry_traced
+    def my_function():
+        ...
+
+    @sentry_traced(transaction_name="new_tx")
+    async def my_async_function():
+        ...
+
+    @sentry_trace(op="child_name")
```
Comment (janfabry) [inline sentry_sdk/decorators.py:40]: Typo: missing `d` at the end of `sentry_trace`.

**Comment by janfabry** [comment]
Would it be better to have a separate decorators for transactions and spans? They have different concerns.

The transaction decorator:
 - Would ensure a transaction exists
 - Would overwrite the name (if passed)
 - Would ensure a passed `sampled` argument is adhered to (if `True` is  passed, but the existing tx is not sampled, it should restart the transaction - because if you start with sampling disabled I think you can't enable it later on, the `SpanRecorder` is not enabled) (calling the decorator `sentry_traced` when it does not ensure it actually is traced is confusing)

The span decorator:
 - Would start a new span, with the passed args
 - Would only work if a transaction exists, but would not care if it doesn't

Maybe you could also have a decorator that combines both of these, but then it needs a clear name.

(I'm struggling with these questions right now as we are designing the decorators for our use case, so maybe it's a useful brain dump, but if not, feel free to ignore it)

File: sentry_sdk/decorators.py (hunk 1 above)
Comment (ynouri) [reply sentry_sdk/decorators.py]: Done

**Comment by ynouri** [comment]
I would be happy to bring this to the finish line but probably best to get feedback from a maintainer first. I moved it from Draft to Ready for Review. Maybe it'll get picked up by someone with this new status. cc @sl0thentr0py

**Comment by antonpirker** [comment]
Hey @ynouri 
I will change the API, ab bit, because we have some things that should be working the same over different programming languages.

Also the span.op should not be set to an arbitrary value because we index spans based on this so we have a fixed set we can choose: https://develop.sentry.dev/sdk/performance/span-operations/

(I will change your code to set it to `function`)

**Comment by antonpirker** [comment]
I have now removed the creation of the transaction in your decorator and also set the span `op` always to `function`. 
I know this might now frustrate you because you did a lot of good work, and now I come along and delete half your code. 
Just be assured, that we really appreciate your contribution, but as Sentry has to support a lot of platforms and versions we have some constraints. Sorry for that.

One thing that we need to fix now: Python 2.7 does not understand the `async def something()` definition of a function. 
I am trying to figure out now, how the decorator can support sync functions in python 2 and 3 and async functions only in python 3.

File: sentry_sdk/tracing.py [hunk 2]
```diff
@@ -803,6 +805,36 @@ def finish(self, hub=None, end_timestamp=None):
...
+    If there is no current transaction, than nothing will be traced.
+
+    Usage:
+        import sentry_sdk
+
+        @sentry_sdk.trace
+        def my_function():
+            ...
+
+        @senrty_sdk.trace
```
Comment (HazAT) [inline sentry_sdk/tracing.py:821]:
```suggestion
        @sentry_sdk.trace
```

**Comment by antonpirker** [comment]
This is the updated documentation: https://github.com/getsentry/sentry-docs/pull/6439

**Comment by antonpirker** [comment]
If you could have a look @sl0thentr0py , would be great. 

There is some duplicated code but this is necessary so the decorator works in Python 2 and in Python 3 with sync and async functions.

File: sentry_sdk/tracing.py [hunk 3]
```diff
@@ -803,6 +805,36 @@ def finish(self, hub=None, end_timestamp=None):
...
+
+        @sentry_sdk.trace
+        def my_function():
+            ...
+
+        @sentry_sdk.trace
+        async def my_async_function():
+            ...
+    """
+    if PY2:
```
Comment (sl0thentr0py) [inline sentry_sdk/tracing.py:825]: style/convention nit but instead of different files can we just switch on the `def` inside `tracing_utils` based on `PY2`, this is the pattern we follow everywhere else.

File: sentry_sdk/tracing_utils.py [hunk 4]
```diff
@@ -395,6 +395,16 @@ def should_propagate_trace(hub, url):
     return False
 
 
+def get_running_span_or_transaction(hub):
```
Comment (sl0thentr0py) [inline sentry_sdk/tracing_utils.py:398]: @antonpirker  we've long wanted to just expose `current_span` and `current_transaction` on the top level api, so maybe split this out into a separate PR and just do it as well.

**Review by sl0thentr0py — CHANGES_REQUESTED** [review]
some convention nits, I really think we don't wanna start splitting files for python 2/3

File: sentry_sdk/tracing.py (hunk 3 above)
Comment (antonpirker) [reply sentry_sdk/tracing.py:825]: This does not work because it is a `async def`. As soon as the Py2 interpreter sees a `async def` in a file it stops parsing it, thas whey the separate file. 
This is the first case we have where we have a async function in code that is also used in python 2. (the `@trace` decorator should work in py2 and py3. 
All the other cases of async code is code that only works in py3 (because those frameworks only support py3)

File: sentry_sdk/tracing_utils.py (hunk 4 above)
Comment (antonpirker) [reply sentry_sdk/tracing_utils.py:398]: Ok, sounds like a useful thing. Have put it here: https://github.com/getsentry/sentry-python/pull/1954

**Comment by sl0thentr0py** [comment]
@antonpirker why are there so many test changes in this PR now? can you summarize them?

**Review by antonpirker — APPROVED** [review]

**Comment by antonpirker** [comment]
Because I now install `pytest_asyncio` for running the common tests some tests (like test_asyncio) have been run for the first time (they where not run at all before) and this lead to some broken tests.

Now I changed the test setup to have a "common" test environment that is kind of like an integration, so I can install `pytest_asyncio` to only this "common" tests without having it also there in all the integration tests.

File: sentry_sdk/tracing_utils_py3.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    """
+
+    # Asynchronous case
+    if inspect.iscoroutinefunction(func):
+
+        @wraps(func)
+        async def func_with_tracing(*args, **kwargs):
+            # type: (*Any, **Any) -> Any
+
+            span_or_trx = get_current_span_or_transaction(sentry_sdk.Hub.current)
```
Comment (sl0thentr0py) [inline sentry_sdk/tracing_utils_py3.py:33]: also as discussed I think `get_current_span` here is sufficient

**Review by sl0thentr0py — CHANGES_REQUESTED** [review]

**Review by sl0thentr0py — APPROVED** [review]

**Comment by antonpirker** [comment]
Hey @ynouri !

Thanks for the great contribution and the patience! Finally we have merged your contribution!

If you want to have a small gift for your work, please send me your shipping address and your t-shirt size to anton.pirker@sentry.io and I will send you something! Thanks!

**Comment by ynouri** [comment]
Thanks @antonpirker for the heavy lifting!
